jerry-ext/jerry-module/include/jerry-module.h

```diff
@@ -0,0 +1,100 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JERRYX_MODULE_H
+#define JERRYX_MODULE_H
+
+#include "jerryscript.h"
+
+#ifdef __cplusplus
+#define EXTERN_C_START extern "C" {
+#define EXTERN_C_END }
+#else
+#define EXTERN_C_START
+#define EXTERN_C_END
+#endif
+
+#define JERRYX_C_CTOR(fn)                          \
+static void fn(void) __attribute__((constructor)); \
+static void fn(void)
+
+#define JERRYX_MODULE_VERSION 1
+
+typedef struct jerryx_module_link
+{
+  jerry_char_t *name;
+  struct jerryx_module_link *next;
+} jerryx_module_header_t;
+
+#define JERRYX_MODULE_HEADER_STATIC_INIT(module_name) \
+  {                                                 \
+    .name = (module_name),                          \
+    .next = NULL                                    \
+  }
+
+#define JERRYX_MODULE_HEADER_UNLINK(link) \
+  ((jerryx_module_header_t *) link)->next = NULL
+
+#define JERRYX_MODULE_HEADER_RUNTIME_INIT(link, module_name) \
+  ((jerryx_module_header_t *) link)->name = (module_name); \
+  JERRYX_MODULE_HEADER_UNLINK((link))
+
+typedef struct
+{
+  jerryx_module_header_t link;
+  int version;
+  jerry_value_t (*init)(void);
+} jerryx_module_t;
+
+/**
+ * Function to create a module manager context
+ */
+void *jerryx_module_context_init (void);
+
+/**
+ * Function to clean up a given module manager context
+ */
+void jerryx_module_context_deinit (void *context_p);
+
+/**
+ * Function to register a module
+ */
+void jerryx_module_register (jerryx_module_t *module);
+
+/**
+ * Function to unregister a module
+ */
+void jerryx_module_unregister (jerryx_module_t *module);
+
+/**
+ * Function to load a copy of a module into the current context
+ */
+jerry_value_t jerryx_module_load (const jerry_char_t *name);
```

Comment:
@gabrielschulhof what platform + compiler are you using?

I just did a quick test on Linux (Ubuntu 14.0.1) and gcc 4.8.4 (both "very" old).

In `main-unix.c` I added in `read_file` (but it doesn't really matter to what function you add it):

```
static uint8_t my_section_test[16] __attribute__((section("my_section"))) __attribute__((used));
extern int __start_my_section;
extern int __stop_my_section;
printf("my_section starts at: %p", &__start_my_section);
printf("my_section ends at: %p", &__stop_my_section);
```

Then build and check `nm`:
```
$./tools/build.py --debug
....
$ nm build/bin/jerry | grep my_section
00000000006d0020 d my_section_test.1742.2506.6498
00000000006d0020 D __start_my_section
00000000006d0030 D __stop_my_section
```
So, for me the "automagic" `__start_...` and `__stop_...` pseudo-symbols are added by GNU ld.
I did not have to add a custom linker script for Linux.

I also tried on macOS (10.12.4) with **gcc 6.3.0.1** and also Apple-clang (the one that ships with Xcode, 8.1.0 (clang-802.0.42)). The section attribute seems to be supported, but I had to write `...((section("__text,my_section")))`. But no "automagic" start/stop symbols unfortunately.
It seems custom linker scripts aren't supported on macOS.
Not a show-stopper in my opinion: when building for macOS, you probably don't care about code size :D and would pick the dynamic/runtime registration alternative.

Bottom line for me: if it works on the Linux that JerryScript CI uses, it's probably good enough to be able to add it to the project. That way we can make sure it doesn't break over time. Non-Linux (embedded) targets potentially need custom linker script foo, but IMO, we can leave that up to the user. And for those targets one can take the route of runtime/dynamic registration as well.